programs/test/selftest.c

```diff
@@ -77,6 +79,105 @@
 #endif
 
 
+#if defined MBEDTLS_SELF_TEST
+/* Sanity check for malloc. This is not expected to fail, and is rather
+ * intended to display potentially useful information about the platform,
+ * in particular the behavior of malloc(0). */
+static int calloc_self_test( int verbose )
+{
+    int failures = 0;
+    void *empty1 = mbedtls_calloc( 0, 1 );
+    void *empty2 = mbedtls_calloc( 0, 1 );
+    void *buffer1 = mbedtls_calloc( 1, 1 );
+    void *buffer2 = mbedtls_calloc( 1, 1 );
+    /* Attempt to allocate a size that's larger than size_t. This
+     * should fail. With AddressSanitizer, this crashes the program by
+     * default; to avoid this, set the environment variable
+     * ASAN_OPTIONS=allocator_may_return_null=1 when running this program.
```

Comment:
On the one hand, I don't like having the self-test program fail when run with ASan.

On the other hand, this is the one aspect of this self-test that might be plausibly broken on an embedded platform with a custom `calloc`, and it wouldn't cause functional bugs but could be a major security issue (e.g. convince the memory allocator that you want to allocate 4 bytes, but make mbedtls code think it's allocated SIZE_MAX+4 bytes).

Reply:
And in fact GCC's warning covers the case where the product of the parameters to `calloc` overflows. So if we perform this test, we need to do it without using values that the compiler can see. Since this is out of scope of this PR, I'll just remove this test.